x-pack/plugins/lens/public/app_plugin/mounter.tsx

```diff
@@ -29,14 +28,17 @@ export async function mountApp(
   createEditorFrame: EditorFrameStart['createInstance']
 ) {
   const [coreStart, startDependencies] = await core.getStartServices();
-  const { data: dataStart, navigation } = startDependencies;
+  const { data: dataStart, navigation, embeddable } = startDependencies;
   const savedObjectsClient = coreStart.savedObjects.client;
   addHelpMenuToAppChrome(coreStart.chrome, coreStart.docLinks);
 
   coreStart.chrome.docTitle.change(
     i18n.translate('xpack.lens.pageTitle', { defaultMessage: 'Lens' })
   );
 
+  const { originatingApp } =
+    embeddable?.stateTransfer.getIncomingOriginatingApp(params.history, true) || {};
```

Comment:
@stacey-gammon, @timroes On the topic of state and `removeAfterFetching`:

I was seeing some strange, inconsistent  behavior with when and how the state was being cleared. In short, dashboard and visualize lost their state on reload, whereas lens did not.

@Dosant was able to diagnose the issue, and we narrowed it down to a conflict between the hash history, used in dashboard and visualize, alongside the browser history. Hash history does not support state, and the browser history is listening to url changes. Because of this, as soon as the url state syncs or changes, anything stored in state is lost. Deangularization of visualize and dashboard should fix this.

here's a [Code sandbox](https://codesandbox.io/s/historyconflict-w565y?file=/src/index.js) that illustrates the problem.

Here's the plan so far:
1. I've used removeAfterFetching here in lens, so that its behavior at least matches visualize and dashboard.
2. After deangularization of visualize and dashboard, we can revisit their usages of stateTransfer so that the intended behavior (i.e. originating app is retained over reloads by lens and visualize, embeddable package is not retained by dashboard to prevent duplicates) is preserved.